File: mybatis-plus/src/test/java/com/baomidou/mybatisplus/test/BaseDbTest.java
package com.baomidou.mybatisplus.test;

import com.baomidou.mybatisplus.core.MybatisConfiguration;
import com.baomidou.mybatisplus.core.MybatisSqlSessionFactoryBuilder;
import com.baomidou.mybatisplus.core.config.GlobalConfig;
import com.baomidou.mybatisplus.core.toolkit.CollectionUtils;
import com.baomidou.mybatisplus.core.toolkit.ExceptionUtils;
import com.baomidou.mybatisplus.core.toolkit.GlobalConfigUtils;
import com.baomidou.mybatisplus.core.toolkit.StringUtils;
import org.apache.ibatis.builder.xml.XMLMapperBuilder;
import org.apache.ibatis.io.Resources;
import org.apache.ibatis.logging.slf4j.Slf4jImpl;
import org.apache.ibatis.mapping.Environment;
import org.apache.ibatis.plugin.Interceptor;
import org.apache.ibatis.session.Configuration;
import org.apache.ibatis.session.ExecutorType;
import org.apache.ibatis.session.SqlSession;
import org.apache.ibatis.session.SqlSessionFactory;
import org.apache.ibatis.transaction.managed.ManagedTransactionFactory;
import org.apache.ibatis.type.TypeReference;
import org.h2.Driver;
import org.jetbrains.annotations.Nullable;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.datasource.SimpleDriverDataSource;

import javax.sql.DataSource;
import java.io.IOException;
import java.io.InputStream;
import java.util.List;
import java.util.function.Consumer;

/**
 * @author miemie
 * @since 2020-06-23
 */
public abstract class BaseDbTest<T> extends TypeReference<T> {

    protected SqlSessionFactory sqlSessionFactory;
    protected Class<T> mapper;
    protected JdbcTemplate jdbcTemplate;

    @SuppressWarnings("unchecked")
    public BaseDbTest() {
        DataSource ds = dataSource();
        List<String> tableSql = tableSql();
        String tableDataSql = tableDataSql();
        String mapperXml = mapperXml();
        GlobalConfig globalConfig = globalConfig();
        List<Interceptor> interceptors = interceptors();
        Consumer<Configuration> consumer = consumer();
        mapper = (Class<T>) getRawType();

        jdbcTemplate = new JdbcTemplate(ds);
        if (CollectionUtils.isNotEmpty(tableSql)) {
            for (String sql : tableSql) {
                if (StringUtils.isNotBlank(sql)) {
                    jdbcTemplate.execute(sql);
                }
            }
        }

        if (StringUtils.isNotBlank(tableDataSql)) {
            jdbcTemplate.execute(tableDataSql);
        }
        MybatisSqlSessionFactoryBuilder builder = new MybatisSqlSessionFactoryBuilder();
        Environment environment = new Environment("test", new ManagedTransactionFactory(), ds);
        MybatisConfiguration configuration = new MybatisConfiguration(environment);
        configuration.setUseDeprecatedExecutor(false);
        if (consumer != null) {
            consumer.accept(configuration);
        }
        GlobalConfigUtils.setGlobalConfig(configuration, globalConfig);
        configuration.setLogImpl(Slf4jImpl.class);
        if (StringUtils.isNotBlank(mapperXml)) {
            try {
                InputStream inputStream = Resources.getResourceAsStream(mapperXml);
                XMLMapperBuilder xmlMapperBuilder = new XMLMapperBuilder(inputStream,
                    configuration, mapperXml, configuration.getSqlFragments());
                xmlMapperBuilder.parse();
            } catch (IOException e) {
                throw ExceptionUtils.mpe(e);
            }
        }
        configuration.addMapper(mapper);
        if (CollectionUtils.isNotEmpty(interceptors)) {
            interceptors.forEach(configuration::addInterceptor);
        }
        sqlSessionFactory = builder.build(configuration);
    }

    private DataSource dataSource() {
        SimpleDriverDataSource dataSource = new SimpleDriverDataSource();
        dataSource.setDriver(new Driver());
        dataSource.setUrl("jdbc:h2:mem:test;MODE=mysql;DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALSE");
        dataSource.setUsername("sa");
        dataSource.setPassword("");
        return dataSource;
    }

    protected SqlSession sqlSession(@Nullable ExecutorType type) {
        return sqlSessionFactory.openSession(type);
    }

    protected void doTest(Consumer<T> consumer) {
        try (SqlSession sqlSession = sqlSession(null)) {
            doTest(sqlSession, consumer);
        }
    }

    protected void doTestAutoCommit(Consumer<T> consumer) {
        try (SqlSession sqlSession = sqlSession(null)) {
            doTestAutoCommit(sqlSession, consumer);
        }
    }

    protected void doTest(SqlSession sqlSession, Consumer<T> consumer) {
        doMapper(sqlSession, false, consumer);
    }

    protected void doTestAutoCommit(SqlSession sqlSession, Consumer<T> consumer) {
        doMapper(sqlSession, true, consumer);
    }

    protected void doMapper(SqlSession sqlSession, boolean commit, Consumer<T> consumer) {
        T t = sqlSession.getMapper(mapper);
        consumer.accept(t);
        if (commit) {
            sqlSession.commit();
        } else {
            sqlSession.rollback();
        }
    }

    protected List<String> tableSql() {
        return null;
    }

    protected String tableDataSql() {
        return null;
    }

    protected String mapperXml() {
        return null;
    }

    protected List<Interceptor> interceptors() {
        return null;
    }

    protected GlobalConfig globalConfig() {
        return GlobalConfigUtils.defaults();
    }

    protected Consumer<Configuration> consumer() {
        return null;
    }
}

File: mybatis-plus-spring/src/test/kotlin/com/baomidou/mybatisplus/test/kotlin/BaseDbTest.kt
package com.baomidou.mybatisplus.test.kotlin

import com.baomidou.mybatisplus.core.MybatisConfiguration
import com.baomidou.mybatisplus.core.MybatisSqlSessionFactoryBuilder
import com.baomidou.mybatisplus.core.MybatisXMLMapperBuilder
import com.baomidou.mybatisplus.core.config.GlobalConfig
import com.baomidou.mybatisplus.core.toolkit.CollectionUtils
import com.baomidou.mybatisplus.core.toolkit.ExceptionUtils
import com.baomidou.mybatisplus.core.toolkit.GlobalConfigUtils
import com.baomidou.mybatisplus.core.toolkit.StringUtils
import com.baomidou.mybatisplus.extension.toolkit.SqlRunner
import org.apache.ibatis.io.Resources
import org.apache.ibatis.logging.slf4j.Slf4jImpl
import org.apache.ibatis.mapping.Environment
import org.apache.ibatis.plugin.Interceptor
import org.apache.ibatis.session.Configuration
import org.apache.ibatis.session.ExecutorType
import org.apache.ibatis.session.SqlSession
import org.apache.ibatis.session.SqlSessionFactory
import org.apache.ibatis.transaction.managed.ManagedTransactionFactory
import org.apache.ibatis.type.TypeReference
import org.h2.Driver
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.jdbc.datasource.SimpleDriverDataSource
import java.io.IOException
import java.util.function.Consumer
import javax.sql.DataSource

/**
 * @author miemie
 * @since 2020-06-23
 */
abstract class BaseDbTest<T> : TypeReference<T>() {
    protected var sqlSessionFactory: SqlSessionFactory
    protected var mapper: Class<T>
    protected var jdbcTemplate: JdbcTemplate

    init {
        SqlRunner.DEFAULT.close()
        val ds = dataSource()
        val tableSql = tableSql()
        val tableDataSql = tableDataSql()
        val mapperXml = mapperXml()
        val globalConfig = globalConfig()
        val interceptors = interceptors()
        val consumer = consumer()
        mapper = rawType as Class<T>
        jdbcTemplate = JdbcTemplate(ds)
        if (CollectionUtils.isNotEmpty(tableSql)) {
            for (sql in tableSql!!) {
                if (StringUtils.isNotBlank(sql)) {
                    jdbcTemplate.execute(sql)
                }
            }
        }
        if (StringUtils.isNotBlank(tableDataSql)) {
            jdbcTemplate.execute(tableDataSql)
        }
        val builder = MybatisSqlSessionFactoryBuilder()
        val environment = Environment("test", ManagedTransactionFactory(), ds)
        val configuration = MybatisConfiguration(environment)
        consumer?.accept(configuration)
        GlobalConfigUtils.setGlobalConfig(configuration, globalConfig)
        configuration.logImpl = Slf4jImpl::class.java
        if (StringUtils.isNotBlank(mapperXml)) {
            try {
                val inputStream = Resources.getResourceAsStream(mapperXml)
                val xmlMapperBuilder = MybatisXMLMapperBuilder(
                    inputStream,
                    configuration, mapperXml, configuration.sqlFragments
                )
                xmlMapperBuilder.parse()
            } catch (e: IOException) {
                throw ExceptionUtils.mpe(e)
            }
        }
        configuration.addMapper(mapper)
        otherMapper().forEach(Consumer { type: Class<*>? -> configuration.addMapper(type) })
        if (CollectionUtils.isNotEmpty(interceptors)) {
            interceptors!!.forEach(Consumer { interceptor: Interceptor? -> configuration.addInterceptor(interceptor) })
        }
        sqlSessionFactory = builder.build(configuration)
    }

    private fun dataSource(): DataSource {
        val dataSource = SimpleDriverDataSource()
        dataSource.driver = Driver()
        dataSource.url = "jdbc:h2:mem:test;MODE=mysql;DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALSE"
        dataSource.username = "sa"
        dataSource.password = ""
        return dataSource
    }

    protected fun sqlSession(type: ExecutorType?): SqlSession {
        return sqlSessionFactory.openSession(type)
    }

    protected fun doTest(consumer: Consumer<T>) {
        sqlSession(null).use { sqlSession -> doTest(sqlSession, consumer) }
    }

    protected fun doTestAutoCommit(consumer: Consumer<T>) {
        sqlSession(null).use { sqlSession -> doTestAutoCommit(sqlSession, consumer) }
    }

    protected fun doTest(sqlSession: SqlSession, consumer: Consumer<T>) {
        doMapper(sqlSession, false, consumer)
    }

    protected fun doTestAutoCommit(sqlSession: SqlSession, consumer: Consumer<T>) {
        doMapper(sqlSession, true, consumer)
    }

    protected fun doMapper(sqlSession: SqlSession, commit: Boolean, consumer: Consumer<T>) {
        val t = sqlSession.getMapper(mapper)
        consumer.accept(t)
        if (commit) {
            sqlSession.commit()
        } else {
            sqlSession.rollback()
        }
    }

    protected open fun tableSql(): List<String?>? {
        return null
    }

    protected open fun tableDataSql(): String? {
        return null
    }

    protected fun mapperXml(): String? {
        return null
    }

    protected fun interceptors(): List<Interceptor?>? {
        return null
    }

    protected fun globalConfig(): GlobalConfig {
        return GlobalConfigUtils.defaults()
    }

    protected fun consumer(): Consumer<Configuration>? {
        return null
    }

    protected fun otherMapper(): List<Class<*>> {
        return emptyList()
    }
}
